### PR TITLE
Add check for Hive 2.0 ACID versioned tables

### PR DESCRIPTION
**Pull request by harmandeeps**
Fixes: #2790

### CODE REVIEW

File: presto-hive/src/main/java/io/prestosql/plugin/hive/BackgroundHiveSplitLoader.java [hunk 1]
```diff
@@ -419,6 +419,18 @@ private void invokeNoMoreSplitsIfNecessary()
...
 
+            if (AcidUtils.isFullAcidTable(table.getParameters())) {
+                /**
+                 * From Hive version >= 3.0, delta/base files will always have file '_orc_acid_version' with value >= '2'.
+                 */
+                Path baseOrDeltaPath = directory.getBaseDirectory() != null ? directory.getBaseDirectory() :
+                        directory.getCurrentDirectories().size() > 0 ? directory.getCurrentDirectories().get(0).getPath() : null;
+
+                if (baseOrDeltaPath != null && AcidUtils.OrcAcidVersion.getAcidVersionFromMetaFile(baseOrDeltaPath, fs) < 2) {
+                    throw new PrestoException(NOT_SUPPORTED, String.format("Hive 2.0 versioned transactional tables are not supported, Please run major compaction in Hive 3.0"));
```
Comment (findepi) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/BackgroundHiveSplitLoader.java]: - `format` is unnecessary. 
- we don't know whether this is Hive 2 or Hive 1. Maybe 
  ```
  "Hive transactional tables are supported with Hive 3.0 and only after a major compaction has been run"
  ```

File: presto-hive/src/test/java/io/prestosql/plugin/hive/TestBackgroundHiveSplitLoader.java [hunk 2]
```diff
@@ -424,8 +429,8 @@ public void testSplitsGenerationWithAbortedTransactions()
         HiveSplitSource hiveSplitSource = hiveSplitSource(backgroundHiveSplitLoader);
         backgroundHiveSplitLoader.start(hiveSplitSource);
         List<String> splits = drain(hiveSplitSource);
-        assertTrue(splits.stream().anyMatch(p -> p.contains(filePaths.get(0))), format("%s not found in splits %s", filePaths.get(0), splits));
-        assertTrue(splits.stream().anyMatch(p -> p.contains(filePaths.get(2))), format("%s not found in splits %s", filePaths.get(2), splits));
+        assertTrue(splits.stream().anyMatch(p -> p.contains(filePaths.get(1))), format("%s not found in splits %s", filePaths.get(0), splits));
```
Comment (findepi) [inline presto-hive/src/test/java/io/prestosql/plugin/hive/TestBackgroundHiveSplitLoader.java]: apparently both `filePaths.get` calls should have the same index here

File: presto-hive/src/test/java/io/prestosql/plugin/hive/TestBackgroundHiveSplitLoader.java [hunk 3]
```diff
@@ -424,8 +429,8 @@ public void testSplitsGenerationWithAbortedTransactions()
...
+        assertTrue(splits.stream().anyMatch(p -> p.contains(filePaths.get(5))), format("%s not found in splits %s", filePaths.get(2), splits));
```
Comment (findepi) [inline presto-hive/src/test/java/io/prestosql/plugin/hive/TestBackgroundHiveSplitLoader.java]: apparently both `filePaths.get` calls should have the same index here

File: presto-hive/src/test/java/io/prestosql/plugin/hive/TestBackgroundHiveSplitLoader.java [hunk 4]
```diff
@@ -443,14 +448,15 @@ public void testFullAcidTableWithOriginalFilesFails()
...
         List<String> filePaths = ImmutableList.of(
                 tablePath + "/000000_1",
+                tablePath + "/delta_0000002_0000002_0000/_orc_acid_version",
                 tablePath + "/delta_0000002_0000002_0000/bucket_00000");
 
         for (String path : filePaths) {
             File file = new File(path);
             if (!file.getParent().equals(tablePath.toString())) {
-                assertTrue(file.getParentFile().mkdirs(), "Failed creating directory " + file.getParentFile());
+                assertTrue(file.getParentFile().exists() || file.getParentFile().mkdirs(), "Failed creating directory " + file.getParentFile());
```
Comment (findepi) [inline presto-hive/src/test/java/io/prestosql/plugin/hive/TestBackgroundHiveSplitLoader.java]: this allows you to remove the `if (!file.getParent().equals(tablePath.toString()))` line above

File: presto-hive/src/test/java/io/prestosql/plugin/hive/TestBackgroundHiveSplitLoader.java [hunk 5]
```diff
@@ -474,6 +480,61 @@ public void testFullAcidTableWithOriginalFilesFails()
...
+                Optional.empty(),
+                ImmutableMap.of("transactional", "true"));
+
+        List<String> filePaths = ImmutableList.of(
+                tablePath + "/000000_1",
+                tablePath + "/delta_0000002_0000002_0000/bucket_00000");
+
+        for (String path : filePaths) {
+            File file = new File(path);
+            if (!file.getParent().equals(tablePath.toString())) {
```
Comment (findepi) [inline presto-hive/src/test/java/io/prestosql/plugin/hive/TestBackgroundHiveSplitLoader.java]: This `if` is redundant

File: presto-hive/src/test/java/io/prestosql/plugin/hive/TestBackgroundHiveSplitLoader.java [hunk 6]
```diff
@@ -474,6 +480,61 @@ public void testFullAcidTableWithOriginalFilesFails()
...
+    {
+        java.nio.file.Path tablePath = Files.createTempDirectory("TestBackgroundHiveSplitLoader");
+        Table table = table(
+                tablePath.toString(),
+                ImmutableList.of(),
+                Optional.empty(),
+                ImmutableMap.of("transactional", "true"));
+
+        List<String> filePaths = ImmutableList.of(
+                tablePath + "/000000_1",
```
Comment (findepi) [inline presto-hive/src/test/java/io/prestosql/plugin/hive/TestBackgroundHiveSplitLoader.java]:
```suggestion
                tablePath + "/000000_1", // _orc_acid_version does not exist so it's assumed to be "ORC ACID version 0"
```

File: presto-hive/src/main/java/io/prestosql/plugin/hive/BackgroundHiveSplitLoader.java [hunk 7]
```diff
@@ -419,6 +419,18 @@ private void invokeNoMoreSplitsIfNecessary()
...
                     true));
 
+            if (AcidUtils.isFullAcidTable(table.getParameters())) {
+                /**
+                 * From Hive version >= 3.0, delta/base files will always have file '_orc_acid_version' with value >= '2'.
+                 */
+                Path baseOrDeltaPath = directory.getBaseDirectory() != null ? directory.getBaseDirectory() :
+                        directory.getCurrentDirectories().size() > 0 ? directory.getCurrentDirectories().get(0).getPath() : null;
+
+                if (baseOrDeltaPath != null && AcidUtils.OrcAcidVersion.getAcidVersionFromMetaFile(baseOrDeltaPath, fs) < 2) {
```
Comment (findepi) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/BackgroundHiveSplitLoader.java]: `getAcidVersionFromMetaFile` call is not wrapped `hdfsEnvironment.doAs` and so i'd expect this to fail in a kerberized env. Yet, the product tests passed. Can you check this gets run with a kerberized HDFS  environment?

Comment (harmandeeps) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/BackgroundHiveSplitLoader.java]: `getAcidVersionFromMetaFile` takes `fs` (FileSystem) object which gets initialized by using `hdfsAuthentication.doAs() `in `hdfsEnvironment.getFileSystem()`. Therefore, it is not failing.

File: presto-hive/src/main/java/io/prestosql/plugin/hive/BackgroundHiveSplitLoader.java [hunk 8]
```diff
@@ -419,6 +419,18 @@ private void invokeNoMoreSplitsIfNecessary()
                     false,
                     true));
 
+            if (AcidUtils.isFullAcidTable(table.getParameters())) {
+                /**
```
[the remark this answers is not in the document]
Comment (harmandeeps) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/BackgroundHiveSplitLoader.java]: fixed.

File: presto-hive/src/main/java/io/prestosql/plugin/hive/BackgroundHiveSplitLoader.java [hunk 9]
```diff
@@ -419,6 +419,18 @@ private void invokeNoMoreSplitsIfNecessary()
...
+                 * From Hive version >= 3.0, delta/base files will always have file '_orc_acid_version' with value >= '2'.
+                 */
+                Path baseOrDeltaPath = directory.getBaseDirectory() != null ? directory.getBaseDirectory() :
+                        directory.getCurrentDirectories().size() > 0 ? directory.getCurrentDirectories().get(0).getPath() : null;
```
[the remark this answers is not in the document]
Comment (harmandeeps) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/BackgroundHiveSplitLoader.java]: fixed.

File: presto-hive/src/main/java/io/prestosql/plugin/hive/BackgroundHiveSplitLoader.java (hunk 1 above)
Comment (harmandeeps) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/BackgroundHiveSplitLoader.java]: Rephrased a little, please take a look.

File: presto-hive/src/test/java/io/prestosql/plugin/hive/TestBackgroundHiveSplitLoader.java (hunk 2 above)
Comment (harmandeeps) [reply presto-hive/src/test/java/io/prestosql/plugin/hive/TestBackgroundHiveSplitLoader.java]: fixed.

File: presto-hive/src/test/java/io/prestosql/plugin/hive/TestBackgroundHiveSplitLoader.java (hunk 3 above)
Comment (harmandeeps) [reply presto-hive/src/test/java/io/prestosql/plugin/hive/TestBackgroundHiveSplitLoader.java]: fixed.

File: presto-hive/src/test/java/io/prestosql/plugin/hive/TestBackgroundHiveSplitLoader.java (hunk 4 above)
Comment (harmandeeps) [reply presto-hive/src/test/java/io/prestosql/plugin/hive/TestBackgroundHiveSplitLoader.java]: fixed.

File: presto-hive/src/test/java/io/prestosql/plugin/hive/TestBackgroundHiveSplitLoader.java (hunk 6 above)
Comment (harmandeeps) [reply presto-hive/src/test/java/io/prestosql/plugin/hive/TestBackgroundHiveSplitLoader.java]: fixed.

File: presto-hive/src/test/java/io/prestosql/plugin/hive/TestBackgroundHiveSplitLoader.java (hunk 5 above)
Comment (harmandeeps) [reply presto-hive/src/test/java/io/prestosql/plugin/hive/TestBackgroundHiveSplitLoader.java]: fixed.

File: presto-hive/src/main/java/io/prestosql/plugin/hive/BackgroundHiveSplitLoader.java (hunk 1 above)
Comment (findepi) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/BackgroundHiveSplitLoader.java]: My proposal was:

> Hive transactional tables are supported with Hive 3.0 and only after a major compaction has been run

Currently it's:

> Hive version <= 2.0 transactional tables are not supported, Please run major compaction in Hive 3.0



The problem with current wording is that exception message should be "declarative", stating the problem.

@electrum can you suggest wording here?

File: presto-hive/src/test/java/io/prestosql/plugin/hive/TestBackgroundHiveSplitLoader.java [hunk 10]
```diff
@@ -474,6 +478,59 @@ public void testFullAcidTableWithOriginalFilesFails()
...
+    }
+
+    private void writeDeltaFileData(File file)
+            throws IOException
+    {
+        if (file.getName().equals("_orc_acid_version")) {
+            Files.write(file.toPath(), "2".getBytes(UTF_8));
+        }
+        else {
+            Files.write(file.toPath(), "test".getBytes(UTF_8));
```
Comment (findepi) [inline presto-hive/src/test/java/io/prestosql/plugin/hive/TestBackgroundHiveSplitLoader.java]: This isn't valid ORC content, so it's better to just use empty file

File: presto-hive/src/test/java/io/prestosql/plugin/hive/TestBackgroundHiveSplitLoader.java [hunk 11]
```diff
@@ -399,14 +400,18 @@ public void testSplitsGenerationWithAbortedTransactions()
...
                 tablePath + "/delta_0000002_0000002_0000/bucket_00000",
+                tablePath + "/delta_0000003_0000003_0000/_orc_acid_version",
                 tablePath + "/delta_0000003_0000003_0000/bucket_00000");
 
         for (String path : filePaths) {
             File file = new File(path);
-            assertTrue(file.getParentFile().mkdirs(), "Failed creating directory " + file.getParentFile());
+            assertTrue(file.getParentFile().exists() || file.getParentFile().mkdirs(), "Failed creating directory " + file.getParentFile());
             assertTrue(file.createNewFile(), "Failed to create file");
+            writeDeltaFileData(file);
```
Comment (findepi) [inline presto-hive/src/test/java/io/prestosql/plugin/hive/TestBackgroundHiveSplitLoader.java]: `createNewFile` on the line above is now redundant

File: presto-hive/src/test/java/io/prestosql/plugin/hive/TestBackgroundHiveSplitLoader.java (hunk 10 above)
Comment (harmandeeps) [reply presto-hive/src/test/java/io/prestosql/plugin/hive/TestBackgroundHiveSplitLoader.java]: fixed.

File: presto-hive/src/test/java/io/prestosql/plugin/hive/TestBackgroundHiveSplitLoader.java (hunk 11 above)
Comment (harmandeeps) [reply presto-hive/src/test/java/io/prestosql/plugin/hive/TestBackgroundHiveSplitLoader.java]: fixed.

File: presto-hive/src/test/java/io/prestosql/plugin/hive/TestBackgroundHiveSplitLoader.java [hunk 12]
```diff
@@ -496,7 +495,7 @@ public void testHive2VersionedFullAcidTableFails()
         for (String path : filePaths) {
             File file = new File(path);
             assertTrue(file.getParentFile().exists() || file.getParentFile().mkdirs(), "Failed creating directory " + file.getParentFile());
-            writeDeltaFileData(file);
+            assertTrue(writeFile(file), "Failed to write file");
```
Comment (findepi) [inline presto-hive/src/test/java/io/prestosql/plugin/hive/TestBackgroundHiveSplitLoader.java]: I liked your previous API better:
```
writeDeltaFileData(file);
```

make the method `void` again and add the assert inside, just for the `file.createNewFile()` case

File: presto-hive/src/test/java/io/prestosql/plugin/hive/TestBackgroundHiveSplitLoader.java [hunk 13]
```diff
@@ -520,15 +519,23 @@ public void testHive2VersionedFullAcidTableFails()
...
+    private boolean writeFile(File file)
     {
-        if (file.getName().equals("_orc_acid_version")) {
-            Files.write(file.toPath(), "2".getBytes(UTF_8));
+        try {
+            if (file.getName().equals("_orc_acid_version")) {
+                Files.write(file.toPath(), "2".getBytes(UTF_8));
+            }
+            else if (file.getName().startsWith("000000_")) {
+                Files.write(file.toPath(), "test".getBytes(UTF_8));
```
Comment (findepi) [inline presto-hive/src/test/java/io/prestosql/plugin/hive/TestBackgroundHiveSplitLoader.java]: I think i don't understand this. Was my comment https://github.com/prestosql/presto/pull/2989#discussion_r388134353 unsound?

I don't understand why "test" is valid content for 000000_ files but not for 000001_
Isn't most files ORC?

**Comment by findepi** [comment]
Merged, thanks!

File: presto-hive/src/test/java/io/prestosql/plugin/hive/TestBackgroundHiveSplitLoader.java (hunk 13 above)
Comment (harmandeeps) [reply presto-hive/src/test/java/io/prestosql/plugin/hive/TestBackgroundHiveSplitLoader.java]: The test `testFullAcidTableWithOriginalFilesFails`  started failing if we don't write anything on original file (000000_). You are correct, this isn't in orc format. I can look into adding ORC content.

File: presto-hive/src/test/java/io/prestosql/plugin/hive/TestBackgroundHiveSplitLoader.java (hunk 12 above)
Comment (harmandeeps) [reply presto-hive/src/test/java/io/prestosql/plugin/hive/TestBackgroundHiveSplitLoader.java]: saw, you already fixed and merged, thank you.

**Pull request closed without merging**